diff --git a/src/packetin_filter/packetin_filter.c b/src/packetin_filter/packetin_filter.c
index 8110c37..2178ffa 100644
--- a/src/packetin_filter/packetin_filter.c
+++ b/src/packetin_filter/packetin_filter.c
@@ -185,6 +185,19 @@ register_dl_type_filter( uint16_t dl_type, uint16_t priority,
 }
 
 
+static void
+register_ip_protocol_filter( uint8_t protocol, uint16_t priority,
+                             const char *service_name, const char *entry_name ) {
+  struct ofp_match ofp_match;
+  memset( &ofp_match, 0, sizeof( struct ofp_match ) );
+  ofp_match.wildcards = OFPFW_ALL & ~( OFPFW_DL_TYPE | OFPFW_NW_PROTO );
+  ofp_match.dl_type = ETH_ETHTYPE_IPV4;
+  ofp_match.nw_proto = protocol;
+
+  insert_match_entry( &ofp_match, priority, service_name, entry_name );
+}
+
+
 // TODO: use pubsub lib.
 static void
 register_any_filter( uint16_t priority, const char *service_name,
@@ -220,6 +233,8 @@ set_match_type( int argc, char *argv[] ) {
     if ( ( service_name = match_type( LLDP_PACKET_IN, argv[ i ] ) ) != NULL ) {
       register_dl_type_filter( ETH_ETHTYPE_LLDP, OFP_DEFAULT_PRIORITY,
                                service_name, "filter-lldp" );
+      register_ip_protocol_filter( 99, OFP_DEFAULT_PRIORITY,
+                                   service_name, "filter-lldp-over-ip" );
     }
     else if ( ( service_name = match_type( ANY_PACKET_IN, argv[ i ] ) ) != NULL ) {
       register_any_filter( 0, service_name, "filter-any" );
